File: tests/protocolo.test.js
'use strict';
/* Um protocolo, quatro casas que o escrevem: o firmware da placa
   (`firmware/src/protocolo.h`), o bridge do robô virtual
   (`bridge/server.js`), a página (`web/rede.js`) e o app
   (`android/.../Traducao.kt`). Duplicação assumida e justificada — quatro
   linguagens, um protocolo — mas duplicação livre para divergir em silêncio.

   Os bytes de cada quadro do firmware são conferidos no
   `tests/protocolo_test.c`. O que se confere aqui é o que nenhum teste de uma
   casa só alcança: que as quatro concordam sobre o número de cada tipo. Um
   `T_DIST` que virasse 0x86 num lado só passaria em todos os testes existentes
   e falharia com a criança na frente do robô. */

const test = require('node:test');
const assert = require('node:assert');
const fs = require('node:fs');
const path = require('node:path');

const RAIZ = path.join(__dirname, '..');
const leia = (p) => fs.readFileSync(path.join(RAIZ, p), 'utf8');

/* A verdade combinada. Mudar um número aqui é mudar o protocolo: quem fizer
   isso tem de mexer nas quatro casas, e é esse o ponto do teste. */
const TIPOS = {
  T_LOAD: 0x01,
  T_RUN: 0x02,
  T_STOP: 0x03,
  T_ARENA: 0x04,     /* só o robô virtual tem arena para trocar */
  T_PC: 0x81,
  T_STATE: 0x82,
  T_TELEM: 0x83,     /* pose; a placa não emite, por não ter física */
  T_VALOR: 0x84,
  T_DIST: 0x85,
};

/* Cada casa declara os seus com uma sintaxe diferente; o que muda é só o que
   vem antes do nome. */
const CASAS = [
  { nome: 'firmware/src/protocolo.h', fonte: 'firmware/src/protocolo.h',
    re: (n) => new RegExp(`#define\\s+${n}\\s+(0x[0-9A-Fa-f]+)`),
    tem: ['T_LOAD', 'T_RUN', 'T_STOP', 'T_PC', 'T_STATE', 'T_VALOR', 'T_DIST'] },
  { nome: 'bridge/server.js', fonte: 'bridge/server.js',
    re: (n) => new RegExp(`${n}\\s*=\\s*(0x[0-9A-Fa-f]+)`),
    tem: ['T_LOAD', 'T_RUN', 'T_STOP', 'T_ARENA', 'T_PC', 'T_STATE', 'T_TELEM',
          'T_VALOR'] },
  { nome: 'web/rede.js', fonte: 'web/rede.js',
    re: (n) => new RegExp(`${n}\\s*=\\s*(0x[0-9A-Fa-f]+)`),
    tem: ['T_LOAD', 'T_RUN', 'T_STOP', 'T_ARENA', 'T_PC', 'T_STATE', 'T_TELEM',
          'T_VALOR', 'T_DIST'] },
  { nome: 'android Traducao.kt',
    fonte: 'android/app/src/main/java/br/educacaocriativa/roboblocos/Traducao.kt',
    re: (n) => new RegExp(`const val ${n}\\s*=\\s*(0x[0-9A-Fa-f]+)`),
    tem: ['T_LOAD', 'T_RUN', 'T_STOP', 'T_ARENA', 'T_PC', 'T_STATE', 'T_TELEM',
          'T_VALOR'] },
];

for (const casa of CASAS) {
  test(`${casa.nome} usa os números combinados do protocolo`, () => {
    const src = leia(casa.fonte);
    for (const nome of casa.tem) {
      const achado = src.match(casa.re(nome));
      assert.ok(achado, `${casa.nome} não declara ${nome}`);
      assert.strictEqual(parseInt(achado[1], 16), TIPOS[nome],
        `${casa.nome}: ${nome} vale ${achado[1]}, e o combinado é ` +
        '0x' + TIPOS[nome].toString(16));
    }
  });
}

test('a placa não sabe emitir pose', () => {
  /* O 0x83 leva x, y e theta. A ESP32 não tem física: pose inventada faria o
     desenho do robô saltar para a origem e a missão se dar por cumprida
     sozinha. O protocolo.h nem declara o tipo, e o protocolo_test.c prova que
     nenhum montador dele o emite. */
  const h = leia('firmware/src/protocolo.h');
  assert.ok(!/#define\s+T_TELEM/.test(h),
    'o firmware não deve declarar T_TELEM');
  assert.ok(/0x83/.test(h),
    'mas o porquê deve estar escrito lá, senão alguém acrescenta de volta');
});

test('o firmware da placa lê o protocolo pelo protocolo.h, e não à mão', () => {
  /* Foi assim que o buraco nasceu: os bytes eram montados soltos dentro do
     main.cpp, onde nenhum teste os alcançava. */
  const main = leia('firmware/src/main.cpp');
  assert.match(main, /#include "protocolo\.h"/);
  assert.ok(!/uint8_t\s+q\[\d\]\s*=\s*\{\s*T_/.test(main),
    'quadro montado à mão no main.cpp: passe pelo protocolo.h');
});

File: tests/protocolo_test.c
/* O buraco que este teste fecha: a tradução do protocolo tem três
   implementações — `bridge/server.js`, `Traducao.kt` e a da placa — e só as
   duas primeiras tinham teste. Se a ordem dos bytes saísse trocada no
   firmware, nada na suíte notaria: o `T_DIST` só é exercitado do lado do
   JavaScript que o lê. Sobrava olho humano na frente do robô.

   Aqui os bytes são conferidos um a um, na ordem em que o DataView do
   navegador os lê (little-endian, ver web/rede.js). */
#include <stdio.h>
#include <string.h>
#include "protocolo.h"

static int falhas;

#define CHECK(cond)                                                        \
    do {                                                                   \
        if (!(cond)) {                                                     \
            printf("  FALHOU %s:%d  %s\n", __FILE__, __LINE__, #cond);     \
            falhas++;                                                      \
        }                                                                  \
    } while (0)

/* ---------- o que a placa manda ---------- */

static void teste_pc_vai_em_little_endian(void) {
    printf("teste_pc_vai_em_little_endian\n");
    uint8_t q[PROTOCOLO_MAX_QUADRO];
    CHECK(protocolo_pc(q, 0x0102) == 3);
    CHECK(q[0] == 0x81);
    CHECK(q[1] == 0x02);      /* byte baixo primeiro */
    CHECK(q[2] == 0x01);
}

static void teste_pc_no_maximo(void) {
    printf("teste_pc_no_maximo\n");
    uint8_t q[PROTOCOLO_MAX_QUADRO];
    protocolo_pc(q, 0xFFFF);
    CHECK(q[1] == 0xFF && q[2] == 0xFF);
    /* O 0xFFFF é o "nenhum pc enviado ainda" do main.cpp; tem de caber. */
    protocolo_pc(q, 0);
    CHECK(q[1] == 0 && q[2] == 0);
}

static void teste_estado_e_sempre_zero_ou_um(void) {
    printf("teste_estado_e_sempre_zero_ou_um\n");
    uint8_t q[PROTOCOLO_MAX_QUADRO];
    CHECK(protocolo_estado(q, 0) == 2);
    CHECK(q[0] == 0x82 && q[1] == 0);
    protocolo_estado(q, 1);
    CHECK(q[1] == 1);
    /* vm.rodando é uint8_t e nada impede um valor esquisito; o navegador
       compara com 1, então normalizar aqui é o que evita "rodando" mudo. */
    protocolo_estado(q, 7);
    CHECK(q[1] == 1);
}

static void teste_valor_positivo(void) {
    printf("teste_valor_positivo\n");
    uint8_t q[PROTOCOLO_MAX_QUADRO];
    CHECK(protocolo_valor(q, 0x01020304) == 5);
    CHECK(q[0] == 0x84);
    CHECK(q[1] == 0x04 && q[2] == 0x03 && q[3] == 0x02 && q[4] == 0x01);
}

static void teste_valor_negativo_vai_em_complemento_de_dois(void) {
    printf("teste_valor_negativo_vai_em_complemento_de_dois\n");
    /* O navegador lê com getInt32(1, true): -1 tem de chegar como FF FF FF FF,
       e não como um zero silencioso. A criança faz 3 - 5 no primeiro dia. */
    uint8_t q[PROTOCOLO_MAX_QUADRO];
    protocolo_valor(q, -1);
    CHECK(q[1] == 0xFF && q[2] == 0xFF && q[3] == 0xFF && q[4] == 0xFF);

    protocolo_valor(q, -2);
    CHECK(q[1] == 0xFE && q[2] == 0xFF && q[3] == 0xFF && q[4] == 0xFF);
}

static void teste_valor_grande_nao_cabe_em_int16(void) {
    printf("teste_valor_grande_nao_cabe_em_int16\n");
    /* 100 × 100 = 10000 cabe; 100 × 1000 não caberia em int16. É a razão de o
       campo ser de 32 bits. */
    uint8_t q[PROTOCOLO_MAX_QUADRO];
    protocolo_valor(q, 100000);
    int32_t volta = (int32_t)((uint32_t)q[1] | ((uint32_t)q[2] << 8) |
                              ((uint32_t)q[3] << 16) | ((uint32_t)q[4] << 24));
    CHECK(volta == 100000);
}

static void teste_distancia(void) {
    printf("teste_distancia\n");
    uint8_t q[PROTOCOLO_MAX_QUADRO];
    CHECK(protocolo_dist(q, 400) == 3);
    CHECK(q[0] == 0x85);
    CHECK(q[1] == 0x90 && q[2] == 0x01);      /* 400 = 0x0190 */
}

static void teste_a_placa_nunca_manda_telemetria(void) {
    printf("teste_a_placa_nunca_manda_telemetria\n");
    /* O 0x83 leva pose, e a placa não tem física. Se um quadro dela saísse com
       esse tipo, o desenho do robô saltaria para a origem e a missão se daria
       por cumprida sozinha. Nenhum montador desta casa pode emitir 0x83. */
    uint8_t q[PROTOCOLO_MAX_QUADRO];
    protocolo_pc(q, 1);       CHECK(q[0] != 0x83);
    protocolo_estado(q, 1);   CHECK(q[0] != 0x83);
    protocolo_valor(q, 1);    CHECK(q[0] != 0x83);
    protocolo_dist(q, 1);     CHECK(q[0] != 0x83);
}

/* ---------- o que a placa recebe ---------- */

static void teste_run_e_stop(void) {
    printf("teste_run_e_stop\n");
    uint8_t run[1] = { 0x02 }, stop[1] = { 0x03 };
    CHECK(protocolo_ler(run, 1).tipo == PEDIDO_RUN);
    CHECK(protocolo_ler(stop, 1).tipo == PEDIDO_STOP);
}

static void teste_load_de_uma_instrucao(void) {
    printf("teste_load_de_uma_instrucao\n");
    uint8_t msg[3 + INSTR_BYTES];
    memset(msg, 0, sizeof(msg));
    msg[0] = 0x01;
    msg[1] = 1;                        /* uma instrução */
    msg[3] = 0xAB;                     /* primeiro byte do programa */

    Pedido p = protocolo_ler(msg, sizeof(msg));
    CHECK(p.tipo == PEDIDO_LOAD);
    CHECK(p.bytes == INSTR_BYTES);
    CHECK(p.programa == msg + 3);
    CHECK(p.programa[0] == 0xAB);
}

static void teste_load_com_contagem_em_dois_bytes(void) {
    printf("teste_load_com_contagem_em_dois_bytes\n");
    /* 256 instruções: 0x0100. É onde a contagem passa a precisar do segundo
       byte, e onde um little-endian trocado apareceria como programa vazio. */
    static uint8_t msg[3 + 256 * INSTR_BYTES];
    memset(msg, 0, sizeof(msg));
    msg[0] = 0x01;
    msg[1] = 0x00;
    msg[2] = 0x01;

    Pedido p = protocolo_ler(msg, sizeof(msg));
    CHECK(p.tipo == PEDIDO_LOAD);
    CHECK(p.bytes == 256 * INSTR_BYTES);
}

static void teste_load_cortado_nao_vira_pedido(void) {
    printf("teste_load_cortado_nao_vira_pedido\n");
    /* Diz que traz duas instruções e traz uma. Carregar isso faria o robô
       andar o começo de algo que a criança não montou. */
    uint8_t msg[3 + INSTR_BYTES];
    memset(msg, 0, sizeof(msg));
    msg[0] = 0x01;
    msg[1] = 2;
    CHECK(protocolo_ler(msg, sizeof(msg)).tipo == PEDIDO_NADA);
}

static void teste_load_com_sobra_nao_vira_pedido(void) {
    printf("teste_load_com_sobra_nao_vira_pedido\n");
    uint8_t msg[3 + INSTR_BYTES + 1];
    memset(msg, 0, sizeof(msg));
    msg[0] = 0x01;
    msg[1] = 1;
    CHECK(protocolo_ler(msg, sizeof(msg)).tipo == PEDIDO_NADA);
}

static void teste_load_maior_que_a_vm_aguenta(void) {
    printf("teste_load_maior_que_a_vm_aguenta\n");
    /* MAX_INSTR + 1 não cabe na VM. A conta n * INSTR_BYTES ainda estoura o
       uint16 lá na frente, então a recusa tem de vir antes. */
    uint8_t msg[3];
    uint32_t instr = MAX_INSTR + 1;
    msg[0] = 0x01;
    msg[1] = (uint8_t)(instr & 0xFF);
    msg[2] = (uint8_t)(instr >> 8);
    CHECK(protocolo_ler(msg, 3).tipo == PEDIDO_NADA);
}

static void teste_load_vazio_e_valido(void) {
    printf("teste_load_vazio_e_valido\n");
    /* Zero instrução é o que chega quando a criança apaga tudo e aperta PLAY:
       carrega um programa vazio, e o robô fica parado. Não é erro. */
    uint8_t msg[3] = { 0x01, 0, 0 };
    Pedido p = protocolo_ler(msg, 3);
    CHECK(p.tipo == PEDIDO_LOAD);
    CHECK(p.bytes == 0);
}

static void teste_lixo_nao_vira_pedido(void) {
    printf("teste_lixo_nao_vira_pedido\n");
    uint8_t arena[1] = { 0x04 };       /* é do robô virtual, não da placa */
    uint8_t nada[1] = { 0x7F };
    CHECK(protocolo_ler(arena, 1).tipo == PEDIDO_NADA);
    CHECK(protocolo_ler(nada, 1).tipo == PEDIDO_NADA);
    CHECK(protocolo_ler(nada, 0).tipo == PEDIDO_NADA);
    CHECK(protocolo_ler(0, 3).tipo == PEDIDO_NADA);
}

static void teste_pedido_sem_load_nao_aponta_para_nada(void) {
    printf("teste_pedido_sem_load_nao_aponta_para_nada\n");
    uint8_t run[1] = { 0x02 };
    Pedido p = protocolo_ler(run, 1);
    CHECK(p.programa == 0 && p.bytes == 0);
}

int main(void) {
    teste_pc_vai_em_little_endian();
    teste_pc_no_maximo();
    teste_estado_e_sempre_zero_ou_um();
    teste_valor_positivo();
    teste_valor_negativo_vai_em_complemento_de_dois();
    teste_valor_grande_nao_cabe_em_int16();
    teste_distancia();
    teste_a_placa_nunca_manda_telemetria();
    teste_run_e_stop();
    teste_load_de_uma_instrucao();
    teste_load_com_contagem_em_dois_bytes();
    teste_load_cortado_nao_vira_pedido();
    teste_load_com_sobra_nao_vira_pedido();
    teste_load_maior_que_a_vm_aguenta();
    teste_load_vazio_e_valido();
    teste_lixo_nao_vira_pedido();
    teste_pedido_sem_load_nao_aponta_para_nada();

    if (falhas) {
        printf("\n%d verificação(ões) falharam\n", falhas);
        return 1;
    }
    printf("\ntodos os testes passaram\n");
    return 0;
}
